### ISSUE TITLE
跑run_pt.sh时，使用--deepspeed ds_zero2_no_offload.json运行卡住了

### ISSUE THREAD

**Issue by guijuzhejiang** [ymcui/Chinese-LLaMA-Alpaca#575]
peft使用的指定的0.3.0dev，运行run_pt.sh时，参数都是默认配置，只修改了模型和token路径，运行后读取数据正常，但卡在下面这里不动了，试了几次都如此，但删除--deepspeed ${deepspeed_config_file}后不会卡住，单报其他错误，感觉问题可能出在deepspeed脚本上，下面是卡住时的log：
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 33/33 [00:12<00:00,  2.66it/s]
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 33/33 [00:12<00:00,  2.65it/s]
[INFO|modeling_utils.py:3283] 2023-06-12 18:58:17,569 >> All model checkpoint weights were used when initializing LlamaForCausalLM.

[INFO|modeling_utils.py:3291] 2023-06-12 18:58:17,569 >> All the weights of LlamaForCausalLM were initialized from the model checkpoint at /media/zzg/GJ_disk01/pretrained_model/text-generation-webui/models/decapoda-research_llama-7b-hf.
If your task is similar to the task the model of the checkpoint was trained on, you can already use LlamaForCausalLM for predictions without further training.
[INFO|configuration_utils.py:537] 2023-06-12 18:58:17,572 >> loading configuration file /media/zzg/GJ_disk01/pretrained_model/text-generation-webui/models/decapoda-research_llama-7b-hf/generation_config.json
[INFO|configuration_utils.py:577] 2023-06-12 18:58:17,572 >> Generate config GenerationConfig {
  "_from_model_config": true,
  "bos_token_id": 0,
  "eos_token_id": 1,
  "pad_token_id": 0,
  "transformers_version": "4.30.0.dev0"
}

06/12/2023 18:58:34 - INFO - __main__ - Init new peft model
06/12/2023 18:58:34 - INFO - __main__ - target_modules: ['q_proj', 'v_proj', 'k_proj', 'o_proj', 'gate_proj', 'down_proj', 'up_proj']
06/12/2023 18:58:34 - INFO - __main__ - lora_rank: 8
trainable params: 429203456 || all params: 6905475072 || trainable%: 6.2154080859739
[INFO|trainer.py:594] 2023-06-12 18:59:44,457 >> max_steps is given, it will override any value given in num_train_epochs
/home/zzg/miniconda3/envs/py39_DL_cu118/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
[2023-06-12 18:59:44,476] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed info: version=0.9.4, git-hash=unknown, git-branch=unknown
06/12/2023 18:59:47 - INFO - torch.distributed.distributed_c10d - Added key: store_based_barrier_key:2 to store for rank: 0
trainable params: 429203456 || all params: 6905475072 || trainable%: 6.2154080859739
/home/zzg/miniconda3/envs/py39_DL_cu118/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
06/12/2023 18:59:51 - INFO - torch.distributed.distributed_c10d - Added key: store_based_barrier_key:2 to store for rank: 1
06/12/2023 18:59:51 - INFO - torch.distributed.distributed_c10d - Rank 1: Completed store-based barrier for key:store_based_barrier_key:2 with 2 nodes.
06/12/2023 18:59:51 - INFO - torch.distributed.distributed_c10d - Rank 0: Completed store-based barrier for key:store_based_barrier_key:2 with 2 nodes.
[2023-06-12 18:59:51,724] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed Flops Profiler Enabled: False
[2023-06-12 18:59:51,728] [INFO] [logging.py:96:log_dist] [Rank 0] Removing param_group that has no 'params' in the client Optimizer
[2023-06-12 18:59:51,728] [INFO] [logging.py:96:log_dist] [Rank 0] Using client Optimizer as basic optimizer
[2023-06-12 18:59:51,761] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed Basic Optimizer = AdamW
[2023-06-12 18:59:51,761] [INFO] [utils.py:54:is_zero_supported_optimizer] Checking ZeRO support for optimizer=AdamW type=<class 'transformers.optimization.AdamW'>
[2023-06-12 18:59:51,761] [WARNING] [engine.py:1116:_do_optimizer_sanity_check] **** You are using ZeRO with an untested optimizer, proceed with caution *****
[2023-06-12 18:59:51,761] [INFO] [logging.py:96:log_dist] [Rank 0] Creating torch.float16 ZeRO stage 2 optimizer
[2023-06-12 18:59:51,761] [INFO] [stage_1_and_2.py:133:__init__] Reduce bucket size 100000000
[2023-06-12 18:59:51,761] [INFO] [stage_1_and_2.py:134:__init__] Allgather bucket size 100000000
[2023-06-12 18:59:51,762] [INFO] [stage_1_and_2.py:135:__init__] CPU Offload: False
[2023-06-12 18:59:51,762] [INFO] [stage_1_and_2.py:136:__init__] Round robin gradient partitioning: False
Using /home/zzg/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...
Using /home/zzg/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...

deepspeed重新install后也同样如此，请问有解决办法吗？

**Comment by airaria** [comment]
用的是几机几卡？有没有试过单机单卡能不能跑通？

**Comment by guijuzhejiang** [comment]
@airaria 用的单击双卡，单击单卡尝试过，也同样卡在这里，下面试运行脚本：
lr=2e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

pretrained_model=/media/GJ_disk01/pretrained_model/text-generation-webui/models/decapoda-research_llama-7b-hf
chinese_tokenizer_path=/media/GJ_disk01/pretrained_model/text-generation-webui/loras/chinese_llama_lora_7b/
dataset_dir=/home/workspace/pycharm/Chinese-LLaMA-Alpaca/data
data_cache=/home/workspace/pycharm/Chinese-LLaMA-Alpaca/data/cache
per_device_train_batch_size=1
per_device_eval_batch_size=1
training_steps=100
gradient_accumulation_steps=1
output_dir=output_dir

deepspeed_config_file=ds_zero2_no_offload.json

torchrun --nnodes 1 --nproc_per_node 2 run_clm_pt_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --data_cache_dir ${data_cache} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --seed $RANDOM \
    --fp16 \
    --max_steps ${training_steps} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.05 \
    --weight_decay 0.01 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --save_steps 500 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --block_size 512 \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --modules_to_save ${modules_to_save} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --gradient_checkpointing \
    --ddp_find_unused_parameters False

**Comment by iMountTai** [comment]
DeepSpeed 安装0.9.2试试

**Comment by guijuzhejiang** [comment]
改了ds==0.9.2,用单击单卡测试，又卡在这里了：
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 33/33 [01:17<00:00,  2.35s/it]
[INFO|modeling_utils.py:3295] 2023-06-13 09:29:06,108 >> All model checkpoint weights were used when initializing LlamaForCausalLM.

[INFO|modeling_utils.py:3303] 2023-06-13 09:29:06,108 >> All the weights of LlamaForCausalLM were initialized from the model checkpoint at /media/zzg/GJ_disk01/pretrained_model/text-generation-webui/models/decapoda-research_llama-7b-hf.
If your task is similar to the task the model of the checkpoint was trained on, you can already use LlamaForCausalLM for predictions without further training.
[INFO|configuration_utils.py:537] 2023-06-13 09:29:06,110 >> loading configuration file /media/zzg/GJ_disk01/pretrained_model/text-generation-webui/models/decapoda-research_llama-7b-hf/generation_config.json
[INFO|configuration_utils.py:577] 2023-06-13 09:29:06,111 >> Generate config GenerationConfig {
  "_from_model_config": true,
  "bos_token_id": 0,
  "eos_token_id": 1,
  "pad_token_id": 0,
  "transformers_version": "4.31.0.dev0"
}

06/13/2023 09:29:23 - INFO - __main__ - Init new peft model
06/13/2023 09:29:23 - INFO - __main__ - target_modules: ['q_proj', 'v_proj', 'k_proj', 'o_proj', 'gate_proj', 'down_proj', 'up_proj']
06/13/2023 09:29:23 - INFO - __main__ - lora_rank: 8
trainable params: 838,418,432 || all params: 7,314,690,048 || trainable%: 11.46211837409628
[INFO|trainer.py:577] 2023-06-13 09:30:33,916 >> max_steps is given, it will override any value given in num_train_epochs
/home/zzg/miniconda3/envs/py39_DL_cu118/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
[2023-06-13 09:30:34,153] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed info: version=0.9.2, git-hash=unknown, git-branch=unknown
06/13/2023 09:30:40 - INFO - torch.distributed.distributed_c10d - Added key: store_based_barrier_key:2 to store for rank: 0
06/13/2023 09:30:40 - INFO - torch.distributed.distributed_c10d - Rank 0: Completed store-based barrier for key:store_based_barrier_key:2 with 1 nodes.
[2023-06-13 09:30:40,749] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed Flops Profiler Enabled: False
[2023-06-13 09:30:40,754] [INFO] [logging.py:96:log_dist] [Rank 0] Removing param_group that has no 'params' in the client Optimizer
[2023-06-13 09:30:40,755] [INFO] [logging.py:96:log_dist] [Rank 0] Using client Optimizer as basic optimizer
[2023-06-13 09:30:40,802] [INFO] [logging.py:96:log_dist] [Rank 0] DeepSpeed Basic Optimizer = AdamW
[2023-06-13 09:30:40,802] [INFO] [utils.py:54:is_zero_supported_optimizer] Checking ZeRO support for optimizer=AdamW type=<class 'transformers.optimization.AdamW'>
[2023-06-13 09:30:40,802] [WARNING] [engine.py:1104:_do_optimizer_sanity_check] **** You are using ZeRO with an untested optimizer, proceed with caution *****
[2023-06-13 09:30:40,802] [INFO] [logging.py:96:log_dist] [Rank 0] Creating torch.float16 ZeRO stage 2 optimizer
[2023-06-13 09:30:40,802] [INFO] [stage_1_and_2.py:133:__init__] Reduce bucket size 100000000
[2023-06-13 09:30:40,802] [INFO] [stage_1_and_2.py:134:__init__] Allgather bucket size 100000000
[2023-06-13 09:30:40,802] [INFO] [stage_1_and_2.py:135:__init__] CPU Offload: False
[2023-06-13 09:30:40,802] [INFO] [stage_1_and_2.py:136:__init__] Round robin gradient partitioning: False
Using /home/zzg/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...

**Comment by guijuzhejiang** [comment]
貌似是系统bug，删除了cache恢复正常，感谢airaria和iMountTai的回复，关闭此问题

**Comment by AliceShen122** [comment]
楼主，请问下你单机双卡，每张卡的显存多少呀？我每张卡16GB，和你执行一样的命令，OOM了。
